server/models/instance-action.js

```diff
@@ -0,0 +1,124 @@
+var debug = require('debug')('MeshModels.server.InstanceAction');
+var path = require('path');
+var util = require('util');
+
+module.exports = function(InstanceAction) {
+  InstanceAction.observe('before save', function(ctx, next) {
+    if (!ctx.instance) return next();
+    var now = Date.now();
+    var action = ctx.instance;
+
+    debug('enter before create: %j', action);
+    action.timestamp = now;
+    action.result = {};
+
+    action.serviceInstance(function(err, instance) {
+      if (err) return next(err);
+      instance.serverService(function(err, service) {
+        if (err) return next(err);
+        instance.executor(function(err, executor) {
+          if (err) return next(err);
+
+          processAction(service, executor, instance, action, next);
+        });
+      });
+    });
+  });
+
+  function processAction(service, executor, instance, action, next) {
+    switch (action.request.sub) {
+      case 'stop-cpu-profiling':
+        runProfile(service, executor, instance, action, 'cpuprofile', next);
+        return this;
+      case 'heap-snapshot':
+        runProfile(service, executor, instance, action, 'heapsnapshot', next);
+        return this;
+      default: {
+        InstanceAction.app.serviceManager.ctlRequest(
+          service,
+          instance,
+          action.request,
+          function(err, res) {
+            if (err) return next(err);
+            action.result = res;
+            next();
+          });
+        return this;
+      }
+    }
+  }
+
+  function runProfile(service, executor, instance, action, type, callback) {
+    if (action.request.target == null) {
+      return callback(Error('Missing required argument: `target`'));
+    }
+
+    var ProfileData = InstanceAction.app.models.ProfileData;
+    var ServerService = InstanceAction.app.models.ServerService;
+
+    var profile = new ProfileData({
+      executorId: executor.id,
+      serverServiceId: service.id,
+      instanceId: instance.id,
+      targetId: action.request.target,
+      startTime: action.timestamp,
+      type: type,
+    });
+
+    profile.save(function(err, profile) {
+      if (err) return callback(err);
+
+      var pathname = [
+        InstanceAction.app.get('restApiRoot'),
+        ServerService.sharedClass.http.path,
+        service.id,
+        ProfileData.sharedClass.http.path,
+        String(profile.id),
+        'download'
+      ].join('/').replace(/\/+/g, '/'); // Compress // to /
```

Comment:
yes, that was my concern. Do you know of any node function that normalizes HTTP paths? perhaps I would url.parse + url.format ?